src/plugins_types/union.c

```diff
@@ -150,6 +150,40 @@ lyb_parse_union(const void *lyb_data, size_t lyb_data_len, uint32_t *type_idx, c
     }
 }
 
+/**
+ * @brief For leafref failures, ensure the appropriate error is propagated, not a type validation failure.
+ *
+ * @param[in] ctx_node Context node for prefix resolution.
+ * @param[in,out] err Error record to be updated
+ * @param[in] type leafref type used to extract target path
+ * @param[in] value value attempting to be stored
+ * @param[in] value_len Length of value that was attempted to be stored.
+ */
+static void
+union_update_lref_err(const struct lyd_node *ctx_node, struct ly_err_item *err, const struct lysc_type *type, const void *value, size_t value_len)
+{
+    const struct lysc_type_leafref *lref;
+    char *valstr = NULL;
+
+    if ((err == NULL) || (type->basetype != LY_TYPE_LEAFREF)) {
+        return;
+    }
+
+    lref = (const struct lysc_type_leafref *)type;
+    free(err->apptag);
+    err->apptag = strdup("instance-required");
+
+    free(err->msg);
+
+    if (lyd_get_value(ctx_node) != NULL) {
+        valstr = strdup(lyd_get_value(ctx_node));
+    } else {
+        valstr = strndup((const char *)value, value_len);
+    }
+    asprintf(&err->msg, LY_ERRMSG_NOLREF_VAL, valstr, lyxp_get_expr(lref->path));
```

Comment:
Missing memory allocation failure checks.

Reply:
done. 

As a side note, I'm not a fan of OOM error checking in general.  On linux at least, it just doesn't work since linux uses an optimistic memory allocation strategy which will return success even if there is no memory available and the kernel will just kill your application when you try to use it.  There's probably exceptions where linux bypasses this optimism on ***huge*** allocations where its relevant (which isn't relevant to libyang), but in general it adds a lot of code that does nothing.

Plus all modern memory-safe languages just abort() on OOM, and likely not everything is C these days....

I even wrote a blog post on it : https://bradhouse.dev/2023/11/how-to-handle-out-of-memory-oom-in-c/

Reply:
I am neither but my previous boss was and so we kept adding the checks everywhere. One argument may be that our libraries are not used only on Linux but on other platforms, some running usually on a very limited HW.